### ISSUE TITLE
Question About Evading SNI filter

### ISSUE THREAD

**Issue by Lanius-collaris** [celzero/rethink-app#921]
I found an interesting tool. [https://codeberg.org/Spilopelia-chinensis/TCB-Teardown-test](https://codeberg.org/Spilopelia-chinensis/TCB-Teardown-test)
When it find a SYN+ACK packet ( not from ::1/128, fc::/7, 127.0.0.0/8, 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16 ), it will send an ACK packet and a RST+ACK packet with an IP TTL low enough to not make it to the receiver ( but high enough to be seen by the monitor ). 

Is it possible to implement this feature without root on Android?

**Comment by ignoramous** [comment]
Yes, but it isn't straight forward to do so. See also: https://github.com/celzero/firestack/issues/1

**Comment by Lanius-collaris** [comment]
> Yes, but it isn't straight forward to do so. See also: [celzero/firestack#1](https://github.com/celzero/firestack/issues/1)

Sorry, I am not familiar with Android. Can TUN device **write outgoing packets** or **read incoming packets**?

**Comment by ignoramous** [comment]
Yep, that's what the TUN devices are for.

**Comment by Lanius-collaris** [comment]
Ah, I remember that the common use of TUN device is **reading outgoing packets** and **writing incoming packets**?

**Comment by ignoramous** [comment]
My bad. Yes, in general, with TUN files: read egress, write ingress :)

But in rdns, you could write egress and read ingress ("emulating a TUN device" may be the right way to think about it) because we run a userspace TCP/IP stack.

Though, I'm curious what the context of asking this question is with evading firewalls, in particular?

**Comment by Lanius-collaris** [comment]
Normal tcp socket can't seem to send a TTL-limited RST before sending data, ClientHello Segmentation does't work when acessing many websites from china (It works when accessing torproject.org). Domain Fronting needs man-in-the-middle attack, is it dangerous? Do you know some methods that don't need privileges?

**Comment by ignoramous** [comment]
Oh, I see. If it requires `root` access, then it wouldn't be possible unless Rethink is set up to emulate the TUN device (like does for turning WireGuard in to a TCP/UDP proxy).

**Comment by Lanius-collaris** [comment]
>Domain Fronting isn't dangerous. In fact, the [ESNI standard](https://blog.cloudflare.com/esni/) is like Domain Fronting, in one sense

In some cases, clients don't use Domain Fronting by themself, but a man-in-the-middle proxy performs Domain Fronting.
\
ClientHello Segmentation (e.g. rethink-app) and sending TTL-limited RST before sending data are transparent (invisible to clients and servers), so I prefer them. The second technique can't seem to be used on Android, I may need to upload the tool to my router. ☹️

**Comment by fortuna** [comment]
You can try to establish a connection, close, then reconnect with the same (src ip, src port, dest ip, dest port).

That will send a FIN to the destination that may confuse the middlebox. It seems to work in Turkmenistan. See **Strategy 3: Free Pass** at https://geneva.cs.umd.edu/papers/2023-web-turkmenistan.pdf
